lib/toml/string.rb

```diff
@@ -10,26 +10,45 @@ def value
     # Replace the unicode escaped characters with the corresponding character
     # e.g. \u03B4 => ?
     def self.decode_unicode(str)
-      str.gsub(/([^\\](?:\\\\)*\\u[\da-f]{4})/i) do |m|
-        m[0...-6] + [m[-4..-1].to_i(16)].pack('U')
+      str.gsub(/\\u([\da-f]{4,6})/i) do |m|
+        [m[2..-1].to_i(16)].pack('U')
       end
     end
 
     # Replace special characters such as line feed and tabs.
     def self.decode_special_char(str)
-      str.gsub(/\\0/, "\0")
-        .gsub(/\\t/, "\t")
-        .gsub(/\\b/, "\b")
-        .gsub(/\\f/, "\f")
-        .gsub(/\\n/, "\n")
-        .gsub(/\\\"/, '"')
-        .gsub(/\\r/, "\r")
+      str.gsub(/\\+(\"|.)/) do |m|
```

Comment:
I think it is simpler and safer to use regular expression `/\\(u[\da-fA-F]{4}|U[\da-fA-F]{6}|.)/` and then just add sequence `'\\\\' to special_chars. And then do decode_unicode inside this block to ensure that all string is scanned with this regular expression just once.